File: game/utilities/HangmanEngine.java
package game.utilities;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;

public class HangmanEngine {

    private final String[] arrayWord;
    private final int wordLength;
    private int guesses;
    private String[] correctGuesses;
    private String incorrectGuesses = "";
    private String allGuesses = "";
    private ArrayList<String> allWordGuesses;
    private String word = "";

    //Constructor
    public HangmanEngine() {
        arrayWord = importWord();
        wordLength = arrayWord.length;
        guesses = 0;
        correctGuesses = arrayHangmanFormatter(wordLength);
        convertArrayedWord();
        this.allWordGuesses = new ArrayList<String>();
    }

    //Accessors
    public int getGuesses() {
        return guesses;
    }
    public int getWordLength() {
        return wordLength;
    }
    public String[] getArrayWord() {
        return arrayWord;
    }
    public String[] getCorrectGuesses() {
        return correctGuesses;
    }
    public String getIncorrectGuesses() {
        return incorrectGuesses;
    }

    //Default Methods
    public void printGameState() {
        for (int i = 0; i < wordLength; i++) {
            System.out.print(correctGuesses[i] + " ");
        }
        System.out.print("   ");
        System.out.print("Used letters: {" + incorrectGuesses + "}\n");
    }

    public boolean processGuess(String guess) {
        if(guess.length() == 1) {
            boolean containsGuess = arraySearcher(arrayWord, guess);
            if (allGuesses.contains(guess)) {
                System.out.println("You have already guessed this letter.");
                return true;
            }
            else {
                if (containsGuess) {
                    correctGuesses = arrayReplacer(arrayWord, correctGuesses, guess);
                    guesses++;
                    allGuesses = allGuesses + guess;
                    return true;
                } else {
                    incorrectGuesses = incorrectGuesses + guess;
                    guesses++;
                    allGuesses = allGuesses + guess;
                    return false;
                }
            }
        }
        else {
            if(allWordGuesses.contains(guess)) {
                System.out.println("You have already guessed this word.");
                return true;
            }
            else {
                if(word.equals(guess)) {
                    for(int i = 0; i < arrayWord.length; i++) {
                        correctGuesses = arrayReplacer(arrayWord, correctGuesses, arrayWord[i]);
                    }
                    guesses++;
                    allWordGuesses.add(guess);
                    return true;
                }
                else {
                    incorrectGuesses = incorrectGuesses + "\"" + guess + "\"";
                    guesses++;
                    allWordGuesses.add(guess);
                    return false;
                }
            }
        }
    }

    public boolean isGuessed() {
        boolean gameOver = true;
        for (int i = 0; i < correctGuesses.length; i++) {
            if (correctGuesses[i].equals("_")) {
                gameOver = false;
            }
        }
        return gameOver;
    }

    public void printPostGameState() {
        if(isGuessed()) {
            System.out.println("YOU WIN!");
            for (int i = 0; i < wordLength; i++) {
                System.out.print(correctGuesses[i] + " ");
            }
            System.out.print("   ");
            System.out.print("Used letters: {" + incorrectGuesses + "}\n");
            System.out.print("You guessed it in " + guesses + " tries.");
        }
        else {
            System.out.println("Better luck next time");
            for (int i = 0; i < wordLength; i++) {
                System.out.print(correctGuesses[i] + " ");
            }
            System.out.print("   ");
            System.out.print("Used guesses: {" + incorrectGuesses + "}\n");
            System.out.println("The word was: " + word);
            System.out.print("You used " + guesses + " tries.");
        }
    }

    //Private Methods
    private static String[] importWord() {
        //Imports 10 words from 'hangmanWords.txt' and chooses one to import to program
        ArrayList<String> allWordsList = new ArrayList<String>();
        String line = "";
        String word = "";
        try {
            //Initializes file reader
            FileReader fr = new FileReader("game/utilities/hangmanWords.txt");
            BufferedReader br = new BufferedReader(fr);

            //Writes words from file to arraylist
            while((line = br.readLine()) != null) {
                allWordsList.add(line);
            }

            br.close();
        }
        catch (IOException ioe) {
            System.out.println("'hangmanWords.txt' Missing");
        }
        //Chooses random word from array and converts to uppercase
        Random rand = new Random();
        int randomNum = rand.nextInt(allWordsList.size()) + 1;
        word = allWordsList.get(randomNum);
        word = word.toUpperCase();
        //Converts chosen word to an array
        String[] arrayedWord = new String[word.length()];
        for (int i = 0; i < word.length(); i++) {
            arrayedWord[i] = Character.toString(word.charAt(i));
        }
        return arrayedWord;
    }
    private static String[] arrayHangmanFormatter(int length) {
        //Fills an array with underscores for Hangman formatting
        String[] array = new String[length];
        for (int i = 0; i < length; i++) {
            array[i] = "_";
        }

        return array;
    }
    private static boolean arraySearcher(String[] array, String guess) {
        //Searches string arrays for given guess
        boolean containsGuess = false;
        if(Arrays.asList(array).contains(guess)) {
            containsGuess = true;
        }
        return containsGuess;
    }
    private static String[] arrayReplacer(String[] word, String[] guessingArray, String replace) {
        //Replaces guessing array with correctly guessed letters
        for(int i = 0; i < word.length; i++) {
            if(word[i].equals(replace)) {
                guessingArray[i] = replace;
            }
        }
        return guessingArray;
    }
    private void convertArrayedWord() {
        for (int i = 0; i < wordLength; i++) {
            word = word + arrayWord[i];
        }
    }

}

File: game/utilities/hangman_engine.c
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <time.h>

int hangman_engine_guesses;
int hangman_engine_incorrect_guesses;
int hangman_engine_word_length;
uint32_t hangman_engine_all_mask;
uint32_t hangman_engine_incorrect_mask;
char *hangman_engine_word;
char *hangman_engine_guess_string;

static void to_upper_case(char *c){
	while(*c){
		if(*c < 'A' || *c > 'Z'){
			*c += 'A' - 'a';
		}
		c++;
	}
}

static void print_guesses(uint32_t guesses){
	char c = 'A';

	while(c <= 'Z'){
		if(guesses&1){
			printf("%c ", c);
		}
		guesses >>= 1;
		c++;
	}
}

static char *import_word(){
	FILE *fp;
	int size;
	char *file_buffer;
	char *output;
	char *c;
	char *start;
	int num_words = 1;
	int rand_word;

	//Read the entire file
	fp = fopen("game/utilities/hangmanWords.txt", "r");
	if(!fp){
		fprintf(stderr, "hangmanWords.txt Missing\n");
		return NULL;
	}

	fseek(fp, 0, SEEK_END);
	size = ftell(fp);
	fseek(fp, 0, SEEK_SET);
	file_buffer = calloc(size + 1, sizeof(char));
	output = calloc(1024, sizeof(char));
	if(!file_buffer || !output){
		fprintf(stderr, "Out of memory\n");
		return NULL;
	}
	fread(file_buffer, sizeof(char), size, fp);
	fclose(fp);

	//Determine the number of words in the words list
	c = file_buffer;
	while(*c){
		if(*c == '\n'){
			num_words++;
		}
		c++;
	}
	
	//Chooses a random word
	rand_word = rand()%num_words;

	//Copies a random word into the output
	c = file_buffer;
	while(rand_word){
		if(*c == '\n'){
			rand_word--;
		}
		c++;
	}
	start = c;
	while(*c && *c != '\n' && c - start < 1023){
		c++;
	}

	memcpy(output, start, c - start);
	free(file_buffer);
	to_upper_case(output);

	return output;
}

static int guess_replace(char *word, char *revealed, char guess){
	int output = 0;

	while(word && *word){
		if(*word == guess){
			*revealed = guess;
			output = 1;
		}
		word++;
		revealed++;
	}

	return output;
}

int hangman_engine_process_guess_char(char guess){
	if(guess < 'A' || guess > 'Z'){
		guess += 'A' - 'a';
	}
	if(hangman_engine_all_mask&(1<<(guess - 'A'))){
		printf("You have already guessed this letter.\n");
		return 1;
	} else {
		hangman_engine_guesses++;
		hangman_engine_all_mask |= 1<<(guess - 'A');
		if(guess_replace(hangman_engine_word, hangman_engine_guess_string, guess)){
			return 1;
		} else {
			hangman_engine_incorrect_guesses++;
			hangman_engine_incorrect_mask |= 1<<(guess - 'A');
			return 0;
		}
	}
}

int hangman_engine_process_guess_string(char *guess){
	to_upper_case(guess);
	if(!strcmp(guess, hangman_engine_word)){
		strcpy(hangman_engine_guess_string, hangman_engine_word);
		hangman_engine_guesses++;
		return 1;
	} else {
		hangman_engine_incorrect_guesses++;
		return 0;
	}
}

void print_post_game_state(){
	if(!strcmp(hangman_engine_word, hangman_engine_guess_string)){
		printf("YOU WIN!\n%s   Used letters: {", hangman_engine_word);
		print_guesses(hangman_engine_incorrect_mask);
		printf("}\nYou guessed it in %d tries.\n", hangman_engine_guesses);
	} else {
		printf("Better luck next time\n%s   Used letters: {", hangman_engine_guess_string);
		print_guesses(hangman_engine_incorrect_mask);
		printf("}\nThe word was: %s\n You used %d tries.\n", hangman_engine_word, hangman_engine_guesses);
	}
}

void print_game_state(){
	printf("%s   Used letters: {", hangman_engine_guess_string);
	print_guesses(hangman_engine_incorrect_mask);
	printf("}\n");
}

void hangman_engine_init(){
	hangman_engine_word = import_word();
	if(!hangman_engine_word){
		exit(1);
	}
	hangman_engine_word_length = strlen(hangman_engine_word);
	hangman_engine_guesses = 0;
	hangman_engine_guess_string = calloc(hangman_engine_word_length + 1, sizeof(char));
	if(!hangman_engine_guess_string){
		fprintf(stderr, "Out of memory\n");
		exit(1);
	}
	memset(hangman_engine_guess_string, '_', hangman_engine_word_length);
	hangman_engine_all_mask = 0;
	hangman_engine_incorrect_mask = 0;
}

void hangman_engine_exit(){
	free(hangman_engine_guess_string);
}
